spring-integration-core/src/main/java/org/springframework/integration/config/MessagingGatewayRegistrar.java

```diff
@@ -163,7 +164,10 @@ else if (StringUtils.hasText(asyncExecutor)) {
 
 		gatewayProxyBuilder.addConstructorArgValue(serviceInterface);
 
-		return new BeanDefinitionHolder(gatewayProxyBuilder.getBeanDefinition(), id);
+		AbstractBeanDefinition beanDefinition = gatewayProxyBuilder.getBeanDefinition();
+		beanDefinition.addMetadataAttribute(new BeanMetadataAttribute("factoryBeanObjectType", serviceInterface));
```

Comment:
Perhaps we should open issues against Spring Framework and Spring Boot to move the constant `FACTORY_BEAN_OBJECT_TYPE` to SF - e.g. to `BeanMetadataAttribute`.

It's a bit brittle to use a literal here and it is possible that other projects may need similar.

For now, I will move this to a constant within SI during the merge.

I will also address @artembilan 's suggestion about fixing the `getObjectType()` when the service interface is null.